main.py

```diff
@@ -877,26 +877,156 @@ def H_ii_indices(det_i: Determinant) -> Iterator[Two_electron_integral_index_pha
-    def category_C(idx, psi_i, psi_j, N_orb):
+    def get_dets_occ_in_orbitals(
+        spindet_occ: Dict[OrbitalIdx, Set[int]],
+        oppspindet_occ: Dict[OrbitalIdx, Set[int]],
+        d_orbitals: Dict[str, Set[OrbitalIdx]],
+        which_orbitals,
+    ):
         """
-        psi_i psi_j: Psi_det, lists of determinants
-        idx: i,j,k,l, index of integral <ij|kl>
-        For an integral i,j,k,l of category C, yield all dependent determinant pairs (I,J)
-        Possibilities are i = k < j < l: (1,2,1,3), i < k < j = l: (1,3,2,3), j < i = k < l: (2,1,2,3)
+        Get indices of determinants that are occupied in the orbitals d_orbitals.
+        Input which_orbitals = "all" or "any" indicates if we want dets occupied in all of the indices, or just any of the indices
         """
-        i, j, k, l = idx
-        (
-            spindet_a_occ_i,
-            spindet_b_occ_i,
-        ) = Hamiltonian_two_electrons_integral_driven.get_spindet_a_occ_spindet_b_occ(psi_i)
+        orbitals_occupied = set()
+        for spintype, indices in d_orbitals.items():
+            if spintype == "same":
+                for o in indices:
+                    if len(orbitals_occupied) == 0:
+                        orbitals_occupied = spindet_occ[o]
+                    else:
+                        if which_orbitals == "all":
+                            orbitals_occupied = orbitals_occupied & spindet_occ[o]
+                        else:
+                            orbitals_occupied = orbitals_occupied | spindet_occ[o]
+            if spintype == "opposite":
+                for o in indices:
+                    if len(orbitals_occupied) == 0:
+                        orbitals_occupied = oppspindet_occ[o]
+                    else:
+                        if which_orbitals == "all":
+                            orbitals_occupied = orbitals_occupied & oppspindet_occ[o]
+                        else:
+                            orbitals_occupied = orbitals_occupied | oppspindet_occ[o]
+        return orbitals_occupied
+
+    @staticmethod
+    def get_dets_via_orbital_occupancy(
+        spindet_occ: Dict[OrbitalIdx, Set[int]],
+        oppspindet_occ: Dict[OrbitalIdx, Set[int]],
+        d_occupied: Dict[str, Set[OrbitalIdx]],
+        d_unoccupied: Dict[str, Set[OrbitalIdx]],
+    ):
+        """
+        get_dets_via_occupency({"same":{i,j}}, {"same":{l}})
+        get_dets_via_occupency({"same":{j}, "opposite"{i}"}, {"same":{l}))
+        """
+
+        return Hamiltonian_two_electrons_integral_driven.get_dets_occ_in_orbitals(
+            spindet_occ, oppspindet_occ, d_occupied, "all"
+        ) - Hamiltonian_two_electrons_integral_driven.get_dets_occ_in_orbitals(
+            spindet_occ, oppspindet_occ, d_unoccupied, "any"
+        )
+
+    @staticmethod
+    def do_diagonal(i, j, k, l, psi_i, spindet_occ_i, oppspindet_occ_i, N_orb):
+        # contribution from integrals to diagonal elements
+        if i != j:  # category B
```

Comment:
Is this strange to have a function specialized from some category, and not just inline them in the category?

We can always create some `do_diagonal` local function in all the category

Reply:
Yeah, we can go either way I think. Here, I was just trying to make a function that works for all the categories it needs to. So splitting it up should be easy if we go that route